cpp/src/arrow/compute/kernels/temporal_internal.h

```diff
@@ -119,19 +152,93 @@ struct ZonedLocalizer {
+    }
+
     try {
-      return zoned_time<Duration>{tz, local_time<Duration>(t)}
+      return zoned_time<Duration>(tz, local_time<Duration>(duration_cast<Duration>(d2)))
           .get_sys_time()
           .time_since_epoch();
-    } catch (const arrow_vendored::date::nonexistent_local_time& e) {
-      *st = Status::Invalid("Local time does not exist: ", e.what());
-      return Duration{0};
-    } catch (const arrow_vendored::date::ambiguous_local_time& e) {
-      *st = Status::Invalid("Local time is ambiguous: ", e.what());
-      return Duration{0};
+    } catch (const arrow_vendored::date::ambiguous_local_time&) {
+      // In case we hit an ambiguous period we round to a time multiple just prior,
+      // convert to UTC and add the time unit we're rounding to.
+      const arrow_vendored::date::local_info li =
+          tz->get_info(local_time<Duration>(duration_cast<Duration>(d2)));
+
+      const Duration t2 =
+          zoned_time<Duration>(
+              tz, local_time<Duration>(duration_cast<Duration>(d2 - li.second.offset)))
+              .get_sys_time()
+              .time_since_epoch();
+      const Unit unit = Unit{options->multiple};
+      const Unit t3 =
+          (t2.count() >= 0) ? t2 / unit * unit : (t2 - unit + Unit{1}) / unit * unit;
+      const Duration t4 = duration_cast<Duration>(t3 + li.first.offset);
+      if (t < t4.count()) {
+        return duration_cast<Duration>(t3 + li.second.offset);
+      }
+      return duration_cast<Duration>(t4);
+    } catch (const arrow_vendored::date::nonexistent_local_time&) {
+      // In case we hit a nonexistent period we calculate the duration between the
+      // start of nonexistent period and rounded to moment in UTC (nonexistent_offset).
+      // We then floor the beginning of the nonexisting period in local time and add
+      // nonexistent_offset to that time point in UTC.
+      const arrow_vendored::date::local_info li =
+          tz->get_info(local_time<Duration>(duration_cast<Duration>(d2)));
+
+      const Duration t2 =
+          zoned_time<Duration>(tz, local_time<Duration>(duration_cast<Duration>(d2)),
+                               arrow_vendored::date::choose::earliest)
+              .get_sys_time()
+              .time_since_epoch();
+      const Duration nonexistent_offset = duration_cast<Duration>(t2 - d2);
+
+      const Unit unit = Unit{options->multiple};
+      const Unit t3 =
+          (t2.count() >= 0) ? t2 / unit * unit : (t2 - unit + Unit{1}) / unit * unit;
+      return duration_cast<Duration>(t3 + li.second.offset) + nonexistent_offset;
+    }
+  }
+
+  template <typename Duration, typename Unit>
+  Duration CeilTimePoint(const int64_t t, const RoundTemporalOptions* options) const {
+    const Duration d = FloorTimePoint<Duration, Unit>(t, options);
+    if (d.count() == t) {
+      return d;
+    }
+    return FloorTimePoint<Duration, Unit>(
+        t + duration_cast<Duration>(Unit{options->multiple}).count(), options);
```

Comment:
It's a bit unexpected to call `floor` twice in the `ceil` implementation, is there perhaps a more efficient way to do this?

Reply:
There definitely is a more efficient way however I think it more logically fits into scope of https://github.com/apache/arrow/pull/13043.